### PR TITLE
Csi test image mirroring

### PR DESCRIPTION
**Pull request by derek-pryor**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds the CSI test container images to `test/utils/image/manifest.go`, so that they can be managed with the tooling that manages test container images for the rest of the e2e tests. Currently the CSI test container images are hard-coded in the different yaml manifest files used by the tests and mirroring these images requires manual work.

The `test/e2e/testing-manifests/storage-csi/update-hostpath.sh` script was updated to:
* Determine the current CSI test image versions that should be used (via inspecting `kubernetes-csi/csi-driver-host-path.git`)
* Find any yaml manifests that contain reference to those images and template them
* (Re-)Generate the `csi-manifest.go` file that is linked into `manifest.go`

The manifest loading code was updated to detect if the yaml file is templated and if so replace the image references with those in `csi-manifest.go`.

#### Special notes for your reviewer:
/sig storage

Image references in the `test/e2e/testing-manifests/storage-csi/gce-pd/` sub-directory are excluded, as they currently contain older image versions then those referenced in `test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml`.

#### Does this PR introduce a user-facing change?

```release-note
Added CSI test container images to `test/utils/image/manifest.go` so that existing test image tooling can now manage CSI test container images as well.
```

### CODE REVIEW

**Comment by dims** [comment]
happy to peek after sig-storage folks sign off! thanks

File: test/e2e/storage/utils/create.go
Comment (jsafrane) [inline test/e2e/storage/utils/create.go:25]: I can see other parts use `text/template`, is there a specific reason for html?

**Comment by jsafrane** [comment]
/assign @pohly 
Can you please check update-hostpath.sh changes?

File: test/e2e/storage/utils/create.go
Comment (derek-pryor) [reply test/e2e/storage/utils/create.go:25]: I think it was an automatic import by my editor...

**Review by pohly — CHANGES_REQUESTED** [review]
I do not fully understand what this PR is trying to achieve.

There is already support for patching image versions: https://github.com/kubernetes/kubernetes/blob/77f6476d346ec0d6a0ed29006415ba5cea4b3f5c/test/e2e/storage/utils/create.go#L624-L636

Isn't this PR doing the same thing? At least the templating approach looks more complicated than patching after loading.

If this somehow changes how images are handled, can you explain what the user-visible difference is?

File: test/utils/image/csi-manifest.go
Comment (pohly) [inline test/utils/image/csi-manifest.go:63]: Instead of generating this file with rather complex shell code, wouldn't it be simpler to just parse the embedded YAML files here and extract the image versions on-the-fly?

Comment (pohly) [reply test/utils/image/csi-manifest.go:63]: I bet that would not only be easier to read and debug, but also less code.

**Comment by derek-pryor** [comment]
The context for this change is that OpenShift wants to better support image mirroring for testing. There is an OpenShift command that can generate a list of e2e test container images to be mirrored that uses the information in `manifest.go`. Currently to support the CSI e2e test container images we have to maintain a separate set of code (in OpenShift) that we have to manually update when there is a new container version that we want to mirror.

The goal of this PR is to add reference to the CSI e2e test container images to the data structure in `manifest.go` so that there is a single source for the CSI test container versions.

@jsafrane did look at several different design options, including your suggestion of extracting the versions from the YAML files. As to why @jsafrane chose the approach of templating the YAML files I have to defer to him. You and him had a short discussion on the CSI slack channel on January 18th about this topic.

File: test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml.in
Comment (pohly) [inline test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml.in:221]: This substitution is based on the assumption that there will be only one hostpath plugin image. Same for the other images.

We may need the flexibility to reference different versions at some point. Right now that occurs for csi-resizer when considering both the hostpath deployment and gcp.

File: test/utils/image/csi-manifest.go
Comment (pohly) [reply test/utils/image/csi-manifest.go:63]: /hold

Please give me a bit more time to explore that approach.

**Comment by jsafrane** [comment]
> There is already support for patching image versions

Patching works, goal of this PR is to have all sidecar + driver images in manifest.go (or anywhere else) where we can list the images easily in go code, so we know *what* images to mirror to a replacement registry. Right now this information is scattered around yaml files.

File: test/utils/image/csi-manifest.go
Comment (pohly) [reply test/utils/image/csi-manifest.go:63]: Please have a look at https://github.com/kubernetes/kubernetes/pull/108458

In contrast to this PR here it's done in pure Go and supports different image versions.

Comment (jsafrane) [reply test/utils/image/csi-manifest.go:63]: That would mean that something in manifest.go would need to have a list of yaml files to parse. The "authoritative" list is deep in the storage e2es and I am afraid of circular deps.

https://github.com/kubernetes/kubernetes/blob/4737127b57d02051dfa9f5f263bcb5049516db1c/test/e2e/storage/drivers/csi.go#L165-L172

Of course, it's either possible to parse every embedded yaml file or have the list public somewhere else.

Comment (pohly) [reply test/utils/image/csi-manifest.go:63]: We can iterate over all YAML files under test/e2e/testing-manifests without having to know how or by which tests those files are used. The resulting list of images might be a bit too broad, but that depends on which tests are going to run. I think it's okay to err on the side of caution and mirror anything that might be used.

**Comment by derek-pryor** [comment]
Closed in favor of https://github.com/kubernetes/kubernetes/pull/108458

**Pull request closed without merging**
